starboard/nplb/posix_compliance/posix_socket_sendmsg_test.cc

```diff
@@ -0,0 +1,366 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace starboard {
+namespace nplb {
+namespace {
+
+TEST(PosixSocketSendmsgTest, SunnyDay) {
+  int sockets[2];
+  ASSERT_EQ(0, socketpair(AF_UNIX, SOCK_STREAM, 0, sockets));
+
+  const char* message = "Hello, world!";
+  struct iovec iov[1];
+  iov[0].iov_base = const_cast<char*>(message);
+  iov[0].iov_len = strlen(message);
+
+  struct msghdr msg = {};
+  msg.msg_iov = iov;
+  msg.msg_iovlen = 1;
+
+  ssize_t bytes_sent = sendmsg(sockets[0], &msg, 0);
+  EXPECT_EQ(strlen(message), static_cast<size_t>(bytes_sent));
+
+  char buffer[32];
+  ssize_t bytes_received = read(sockets[1], buffer, sizeof(buffer));
+  EXPECT_EQ(bytes_sent, bytes_received);
+  EXPECT_EQ(0, strncmp(message, buffer, bytes_received));
+
+  EXPECT_TRUE(close(sockets[0]) == 0);
+  EXPECT_TRUE(close(sockets[1]) == 0);
+}
+
+TEST(PosixSocketSendmsgTest, SunnyDayUdpWithRecvmsg) {
+  int server_fd = socket(AF_INET, SOCK_DGRAM, 0);
+  ASSERT_NE(-1, server_fd);
+
+  struct sockaddr_in server_addr = {};
+  server_addr.sin_family = AF_INET;
+  server_addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+  server_addr.sin_port = 0;  // Bind to any available port
+  ASSERT_EQ(
+      0, bind(server_fd, (struct sockaddr*)&server_addr, sizeof(server_addr)));
+
+  socklen_t server_addr_len = sizeof(server_addr);
+  ASSERT_EQ(0, getsockname(server_fd, (struct sockaddr*)&server_addr,
+                           &server_addr_len));
+
+  int client_fd = socket(AF_INET, SOCK_DGRAM, 0);
+  ASSERT_NE(-1, client_fd);
+
+  struct sockaddr_in client_addr = {};
+  client_addr.sin_family = AF_INET;
+  client_addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+  client_addr.sin_port = 0;  // Bind to any available port
+  ASSERT_EQ(
+      0, bind(client_fd, (struct sockaddr*)&client_addr, sizeof(client_addr)));
+
+  socklen_t client_addr_len = sizeof(client_addr);
+  ASSERT_EQ(0, getsockname(client_fd, (struct sockaddr*)&client_addr,
+                           &client_addr_len));
+
+  const char* message = "Hello, UDP!";
+  struct iovec iov[1];
+  iov[0].iov_base = const_cast<char*>(message);
+  iov[0].iov_len = strlen(message);
+
+  struct msghdr send_msg = {};
+  send_msg.msg_name = &server_addr;
+  send_msg.msg_namelen = sizeof(server_addr);
+  send_msg.msg_iov = iov;
+  send_msg.msg_iovlen = 1;
+
+  ssize_t bytes_sent = sendmsg(client_fd, &send_msg, 0);
+  EXPECT_EQ(strlen(message), static_cast<size_t>(bytes_sent));
+
+  char buffer[32];
```

Comment:
We should address the buffer size problem here and below with whatever we do above